File: repertoire/concepts/data-integration/providers/quality_rule/type_check.ts
// Quality Rule Provider: Type Check Validation
// Validates that field values match their declared type.
// Dimension: validity

export const PROVIDER_ID = 'type_check';
export const PLUGIN_TYPE = 'quality_rule';

export interface FieldDef {
  name: string;
  type: string;
  required?: boolean;
  constraints?: Record<string, unknown>;
}

export interface RuleConfig {
  options?: Record<string, unknown>;
  threshold?: number;
}

export interface RuleResult {
  valid: boolean;
  message?: string;
  severity: 'error' | 'warning' | 'info';
}

export type QualityDimension = 'completeness' | 'uniqueness' | 'validity' | 'consistency' | 'timeliness' | 'accuracy';

export class TypeCheckQualityProvider {
  validate(
    value: unknown,
    field: FieldDef,
    _record: Record<string, unknown>,
    config: RuleConfig
  ): RuleResult {
    if (value === null || value === undefined) {
      return { valid: true, severity: 'error' };
    }

    const strict = config.options?.strict === true;
    const declaredType = field.type.toLowerCase();
    const result = this.checkType(value, declaredType, strict);

    if (!result.valid) {
      return {
        valid: false,
        message: `Field '${field.name}' expected type '${declaredType}' but received incompatible value.`,
        severity: 'error',
      };
    }

    return { valid: true, severity: 'error' };
  }

  private checkType(value: unknown, declaredType: string, strict: boolean): { valid: boolean } {
    switch (declaredType) {
      case 'string':
        if (strict) {
          return { valid: typeof value === 'string' };
        }
        return { valid: typeof value === 'string' || typeof value === 'number' || typeof value === 'boolean' };

      case 'number':
      case 'integer':
      case 'float':
        if (typeof value === 'number') {
          if (declaredType === 'integer' && !Number.isInteger(value)) {
            return { valid: !strict };
          }
          return { valid: !isNaN(value) };
        }
        if (!strict && typeof value === 'string') {
          const parsed = Number(value);
          if (!isNaN(parsed)) {
            if (declaredType === 'integer') {
              return { valid: Number.isInteger(parsed) };
            }
            return { valid: true };
          }
        }
        return { valid: false };

      case 'boolean':
        if (typeof value === 'boolean') return { valid: true };
        if (!strict && typeof value === 'string') {
          return { valid: value === 'true' || value === 'false' };
        }
        return { valid: false };

      case 'date':
      case 'datetime':
        if (value instanceof Date) {
          return { valid: !isNaN(value.getTime()) };
        }
        if (typeof value === 'string') {
          const parsed = Date.parse(value);
          return { valid: !isNaN(parsed) };
        }
        return { valid: false };

      case 'array':
        return { valid: Array.isArray(value) };

      case 'object':
        return { valid: typeof value === 'object' && !Array.isArray(value) && value !== null };

      default:
        return { valid: true };
    }
  }

  appliesTo(_field: FieldDef): boolean {
    return true;
  }

  dimension(): QualityDimension {
    return 'validity';
  }
}

export default TypeCheckQualityProvider;

File: repertoire/concepts/data-integration/providers/quality_rule/TypeCheck.swift
// Quality Rule Provider: Type Check Validation
// Validates that field values match their declared type.
// Dimension: validity

import Foundation

public final class TypeCheckQualityProvider {
    public init() {}

    public func validate(value: Any?, field: FieldDef, record: [String: Any], config: RuleConfig) -> RuleResult {
        guard let value = value, !(value is NSNull) else {
            return RuleResult(valid: true, message: nil, severity: "error")
        }

        let strict = config.options?["strict"] as? Bool ?? false
        let declaredType = field.type_.lowercased()

        if !checkType(value: value, declaredType: declaredType, strict: strict) {
            return RuleResult(
                valid: false,
                message: "Field '\(field.name)' expected type '\(declaredType)' but received incompatible value.",
                severity: "error"
            )
        }

        return RuleResult(valid: true, message: nil, severity: "error")
    }

    private func checkType(value: Any, declaredType: String, strict: Bool) -> Bool {
        switch declaredType {
        case "string":
            if value is String { return true }
            if !strict { return value is Int || value is Double || value is Bool }
            return false

        case "number", "float":
            if value is Int || value is Double { return true }
            if !strict, let s = value as? String, Double(s) != nil { return true }
            return false

        case "integer":
            if value is Int { return true }
            if let d = value as? Double {
                if strict { return false }
                return d.truncatingRemainder(dividingBy: 1) == 0
            }
            if !strict, let s = value as? String, Int(s) != nil { return true }
            return false

        case "boolean":
            if value is Bool { return true }
            if !strict, let s = value as? String {
                return s == "true" || s == "false"
            }
            return false

        case "date", "datetime":
            if value is Date { return true }
            if let s = value as? String {
                let formatter = ISO8601DateFormatter()
                formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
                if formatter.date(from: s) != nil { return true }
                // Try simpler date format
                let simpleFormatter = DateFormatter()
                simpleFormatter.dateFormat = "yyyy-MM-dd"
                if simpleFormatter.date(from: s) != nil { return true }
            }
            return false

        case "array":
            return value is [Any]

        case "object":
            return value is [String: Any]

        default:
            return true
        }
    }

    public func appliesTo(field: FieldDef) -> Bool {
        return true
    }

    public func dimension() -> QualityDimension {
        return .validity
    }
}
